### DISCUSSION TITLE
Ideas for Container-Hosting instead of Web-Hosting?

### DISCUSSION THREAD

**Discussion by lab-at-nohl** [tine-groupware/tine#40] [General]
Can Who can recommend a Hoster or an easy setup (which Hoster?) to run _Tine Groupware_ **as a Container**?

I'd like to discuss possibilities for those who do not run a full Linux-Server anyway. With responsibility of someone else to run basic Services (and do backups) professionally. I know you can stack everything together diy; But it should be as simple as Web-Hosting. The following Services are required: 

- Multi Domain management etc. 
- The MySQL/MariaDB Database Server
- A full SMTP/IMAP mail server
- A proxy or load balancer like nginx including Let's Encrypt certificates (although Tine's Docker Container comes with nginx, you probably need something "on the front"). 

Ideally, the aforementioned Services can be configured by a Web interface (GUI).

**Comment by msphn** [comment]
Believing containers are always a high-end solution and nothing casual. I do run and maintain some containerised applications in AWS and even though, in a mostly managed, environment it takes time to maintain, setup and monitor.

Any setup that allows to spawn those bigger environments including mail service, dns, database, redis are always custom solutions. 

If you want something mostly simple, you could use https://www.froxlor.org/ on setup any VPS with services you have mentioned and at least using templates and some solid default configuration to make that work. This comes closed to what you imagine. If you want a second tine, you would just create a new hosting in there.

In AWS, I use AppRunner for single containers, there comes a problem since PHP doesn't run well in single container environments because you need multiple services, like NGINX and FPM and not just a single foreground process. That's not pretty, but it does indeed work.

If you want to go the AWS path and container, it would do following:

1. AppRunner: with a container with nginx+fpm and supervisord as foreground process (AppRunner does also provide SSL, Load Balancer and Scaling for that one container!)
2. RDS: For the database
3. Route53: For domain and routing to the AppRunner instace
4. WorkMail: As mail service, alternatively any external mail hoster

You could also do that pretty much automatically to set up using infrastructure as code in AWS with cloud formation.

Dunno if that helps you, it's just my thoughts.

**Comment by lab-at-nohl** [comment]
Thanks @mspahn for the insights. I personally do not have any experience in the AWS world. Very much appreciated. Especially the list of Amazon's Apps for the services Tine can connect to. 

I have tried a different way during the last month: **Opensuse Leap _Micro_**. As the name suggests it is a small/[lightweighed distro for running containers](https://get.opensuse.org/leapmicro/5.5/). Also it is one of the newer _immutable_ distros which you probably can destroy but restore the previous situation by a click. To start you own (virtual) server this distro feels just right (it takes you three minutes literally). 

However, it comes with a very limited amount of packages. Thus, you definitely need to run containers - for every little tool. Suse prefers _podman_ before docker but both use identical commands, no learning needed. For beginners you can orchestrate by a **Web UI** (cockpit). Nevertheless, more complex stuff e. g. running containers rootless  and startup them at boot requires shell access. I learned the hard way to miss docker-compose, at latest when an update of a quite standard container became necessary. You never remember what you have typed into a UI. But you can preconfigure some stuff by so-called pods (if you really know what you need and if you do not want to change settings later). This is kind of a container around containers to share network, mounts etc. 

I wonder that it comes with really few useful stuff. It would be nice if e.g. _Traefik_ could be installed and administrated by cockpit. And also LDAP and mail. 

Not as easy as webhosting. But to get started it takes no time EASILY.
